File: app/src/main/java/com/xnote/app/domain/rules/MarkdownConversionRules.kt
package com.xnote.app.domain.rules

import com.xnote.app.domain.document.DrawingBlock
import com.xnote.app.domain.document.ImageBlock
import com.xnote.app.domain.document.NoteDocument
import com.xnote.app.domain.document.StickerBlock
import com.xnote.app.domain.document.TableBlock
import com.xnote.app.domain.document.TextBlock
import com.xnote.app.domain.model.Note
import com.xnote.app.domain.model.NoteKind

// -- Type Definitions

enum class ConversionBlocker {
    AlreadyMarkdown,
    Image,
    Sticker,
    Drawing,
}

// -- Functions

fun conversionBlockers(document: NoteDocument): Set<ConversionBlocker> {
    val blockers = linkedSetOf<ConversionBlocker>()
    for (block in document.blocks) {
        when (block) {
            is ImageBlock -> blockers += ConversionBlocker.Image
            is StickerBlock -> blockers += ConversionBlocker.Sticker
            is DrawingBlock -> blockers += ConversionBlocker.Drawing
            is TextBlock, is TableBlock -> Unit
        }
    }
    return blockers
}

fun conversionBlockers(note: Note): Set<ConversionBlocker> {
    if (note.kind == NoteKind.Markdown) {
        return setOf(ConversionBlocker.AlreadyMarkdown)
    }
    return conversionBlockers(note.document ?: NoteDocument())
}

fun canConvertToMarkdown(note: Note): Boolean = conversionBlockers(note).isEmpty()

File: flutter/lib/domain/rules/markdown_conversion_rules.dart
import '../../core/ids/id_generator.dart';
import '../../core/time/clock.dart';
import '../document/note_block.dart';
import '../document/note_document.dart';
import '../markdown/rich_note_markdown.dart';
import '../model/note.dart';
import '../model/note_revision.dart';
import '../text/note_plain_text.dart';

// -- Type Definitions

enum ConversionBlocker { alreadyMarkdown, image, sticker, drawing }

final class MarkdownConversion {
  const MarkdownConversion({required this.revision, required this.note});

  final NoteRevision revision;
  final Note note;
}

// -- Functions

Set<ConversionBlocker> documentConversionBlockers(NoteDocument document) {
  final blockers = <ConversionBlocker>{};
  for (final block in document.blocks) {
    switch (block) {
      case ImageBlock():
        blockers.add(ConversionBlocker.image);
      case StickerBlock():
        blockers.add(ConversionBlocker.sticker);
      case DrawingBlock():
        blockers.add(ConversionBlocker.drawing);
      case TextBlock() || TableBlock():
        break;
    }
  }
  return Set<ConversionBlocker>.unmodifiable(blockers);
}

Set<ConversionBlocker> noteConversionBlockers(Note note) {
  if (note.kind == NoteKind.markdown) {
    return const <ConversionBlocker>{ConversionBlocker.alreadyMarkdown};
  }
  return documentConversionBlockers(note.document!);
}

bool canConvertToMarkdown(Note note) => noteConversionBlockers(note).isEmpty;

MarkdownConversion convertRichNoteToMarkdown(
  Note note, {
  required IdGenerator idGenerator,
  required Clock clock,
}) {
  final blockers = noteConversionBlockers(note);
  if (blockers.isNotEmpty) {
    throw StateError('Note cannot be converted to Markdown: $blockers');
  }
  final now = clock.nowEpochMilliseconds();
  final revision = NoteRevision(
    id: idGenerator.nextId(),
    noteId: note.id,
    reason: RevisionReason.convertToMarkdown,
    kind: note.kind,
    title: note.title,
    document: note.document,
    markdownText: note.markdownText,
    createdAtEpochMilliseconds: now,
  );
  final markdown = richNoteToMarkdown(note.title, note.document!);
  final converted = withDerivedText(
    note.copyWith(
      kind: NoteKind.markdown,
      document: null,
      markdownText: markdown,
      updatedAtEpochMilliseconds: now,
    ),
  );
  return MarkdownConversion(revision: revision, note: converted);
}
